# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## DataFrame Basics III

### Exercise 8: apply(), MultiIndex and Strings

In [1]:
# run the cell!
import pandas as pd

In [2]:
# run the cell!
cars = pd.read_csv("cars.csv")

In [3]:
# run the cell!
cars

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
5    15.0          8         429.0       198.0    4341          10.0   
6    14.0          8         454.0       220.0    4354           9.0   
7    14.0          8         440.0       215.0    4312           8.5   
8    14.0          8         455.0       225.0    4425          10.0   
9    15.0          8         390.0       190.0    3850           8.5   
10   15.0          8         383.0       170.0    3563          10.0   
11   14.0          8         340.0       160.0    3609           8.0   
12   15.0          8         400.0       150.0    3761           9.5   
13   14.0          8         455.0       225.0    3086          10.0   
14   24.0          4         113.0        95.0    2372          15.0   
15   22.0          6         198.0        95.0    2833          15.5   
16   18.0          6         199.0        97.0    2774          15.5   
17   21.0          6         200.0        85.0    2587          16.0   
18   27.0          4          97.0        88.0    2130          14.5   
19   26.0          4          97.0        46.0    1835          20.5   
20   25.0          4         110.0        87.0    2672          17.5   
21   24.0          4         107.0        90.0    2430          14.5   
22   25.0          4         104.0        95.0    2375          17.5   
23   26.0          4         121.0       113.0    2234          12.5   
24   21.0          6         199.0        90.0    2648          15.0   
25   10.0          8         360.0       215.0    4615          14.0   
26   10.0          8         307.0       200.0    4376          15.0   
27   11.0          8         318.0       210.0    4382          13.5   
28    9.0          8         304.0       193.0    4732          18.5   
29   27.0          4          97.0        88.0    2130          14.5   
..    ...        ...           ...         ...     ...           ...   
368  27.0          4         112.0        88.0    2640          18.6   
369  34.0          4         112.0        88.0    2395          18.0   
370  31.0          4         112.0        85.0    2575          16.2   
371  29.0          4         135.0        84.0    2525          16.0   
372  27.0          4         151.0        90.0    2735          18.0   
373  24.0          4         140.0        92.0    2865          16.4   
374  23.0          4         151.0         NaN    3035          20.5   
375  36.0          4         105.0        74.0    1980          15.3   
376  37.0          4          91.0        68.0    2025          18.2   
377  31.0          4          91.0        68.0    1970          17.6   
378  38.0          4         105.0        63.0    2125          14.7   
379  36.0          4          98.0        70.0    2125          17.3   
380  36.0          4         120.0        88.0    2160          14.5   
381  36.0          4         107.0        75.0    2205          14.5   
382  34.0          4         108.0        70.0    2245          16.9   
383  38.0          4          91.0        67.0    1965          15.0   
384  32.0          4          91.0        67.0    1965          15.7   
385  38.0          4          91.0        67.0    1995          16.2   
386  25.0          6         181.0       110.0    2945          16.4   
387  38.0          6         262.0        85.0    3015          17.0   
388  26.0          4         156.0        92.0    2585          14.5   
389  22.0          6         232.0       112.0    2835          14.7   
390  32.0          4         144.0        96.0    2665          13.9   
391  36.0          4         135.0        84.0

61. __Calculate__ for all __numerical columns__ the __range between highest and lowest__ value! __Fill in the gaps!__ The range for __mpg__ is...?

In [5]:
#numerical columns

cars.loc[:,['cylinders','displacement','horsepower','weight','acceleration']]

cylinders  displacement  horsepower  weight  acceleration
0            8         307.0       130.0    3504          12.0
1            8         350.0       165.0    3693          11.5
2            8         318.0       150.0    3436          11.0
3            8         304.0       150.0    3433          12.0
4            8         302.0       140.0    3449          10.5
5            8         429.0       198.0    4341          10.0
6            8         454.0       220.0    4354           9.0
7            8         440.0       215.0    4312           8.5
8            8         455.0       225.0    4425          10.0
9            8         390.0       190.0    3850           8.5
10           8         383.0       170.0    3563          10.0
11           8         340.0       160.0    3609           8.0
12           8         400.0       150.0    3761           9.5
13           8         455.0       225.0    3086          10.0
14           4         113.0        95.0    2372          15.0
15           6         198.0        95.0    2833          15.5
16           6         199.0        97.0    2774          15.5
17           6         200.0        85.0    2587          16.0
18           4          97.0        88.0    2130          14.5
19           4          97.0        46.0    1835          20.5
20           4         110.0        87.0    2672          17.5
21           4         107.0        90.0    2430          14.5
22           4         104.0        95.0    2375          17.5
23           4         121.0       113.0    2234          12.5
24           6         199.0        90.0    2648          15.0
25           8         360.0       215.0    4615          14.0
26           8         307.0       200.0    4376          15.0
27           8         318.0       210.0    4382          13.5
28           8         304.0       193.0    4732          18.5
29           4          97.0        88.0    2130          14.5
..         ...           ...         ...     ...           ...
368          4         112.0        88.0    2640          18.6
369          4         112.0        88.0    2395          18.0
370          4         112.0        85.0    2575          16.2
371          4         135.0        84.0    2525          16.0
372          4         151.0        90.0    2735          18.0
373          4         140.0        92.0    2865          16.4
374          4         151.0         NaN    3035          20.5
375          4         105.0        74.0    1980          15.3
376          4          91.0        68.0    2025          18.2
377          4          91.0        68.0    1970          17.6
378          4         105.0        63.0    2125          14.7
379          4          98.0        70.0    2125          17.3
380          4         120.0        88.0    2160          14.5
381          4         107.0        75.0    2205          14.5
382          4         108.0        70.0    2245          16.9
383          4          91.0        67.0    1965          15.0
384          4          91.0        67.0    1965          15.7
385          4          91.0        67.0    1995          16.2
386          6         181.0       110.0    2945          16.4
387          6         262.0        85.0    3015          17.0
388          4         156.0        92.0    2585          14.5
389          6         232.0       112.0    2835          14.7
390          4         144.0        96.0    2665          13.9
391          4         135.0        84.0    2370          13.0
392          4         151.0        90.0    2950          17.3
393          4         140.0        86.0    2790          15.6
394          4          97.0        52.0    2130          24.6
395          4         135.0        84.0    2295          11.6
396          4         120.0        79.0    2625          18.6
397          4         119.0        82.0    2720          19.4

[398 rows x 5 columns]

In [13]:
cars.loc[:,['mpg','cylinders','displacement','horsepower','weight','acceleration']].apply(lambda x:x.max()-x.min(),axis=0)

mpg               37.6
cylinders          5.0
displacement     387.0
horsepower       184.0
weight          3527.0
acceleration      16.8
dtype: float64

The range for mpg is ... 37.6!

62. Create a __MultiIndex__ for the __cars__ DataFrame with the columns __"model_year"__ (outer level) and __"origin"__ (inner level). __Save__ the change!

In [113]:
cars.set_index(['model_year','origin'],inplace=True)

63. __Sort__ both index levels in __ascending__ order!. __Save__ the change!

In [114]:
cars.sort_index(ascending=True,inplace=True)

__Inspect__ Cars! What is the __name__ of the __second__ car?

In [115]:
# run the cell!
cars

mpg  cylinders  displacement  horsepower  weight  \
model_year origin                                                      
70         Europe  26.0          4          97.0        46.0    1835   
           Europe  25.0          4         110.0        87.0    2672   
           Europe  24.0          4         107.0        90.0    2430   
           Europe  25.0          4         104.0        95.0    2375   
           Europe  26.0          4         121.0       113.0    2234   
           Japan   24.0          4         113.0        95.0    2372   
           Japan   27.0          4          97.0        88.0    2130   
           Usa     18.0          8         307.0       130.0    3504   
           Usa     15.0          8         350.0       165.0    3693   
           Usa     18.0          8         318.0       150.0    3436   
           Usa     16.0          8         304.0       150.0    3433   
           Usa     17.0          8         302.0       140.0    3449   
           Usa     15.0          8         429.0       198.0    4341   
           Usa     14.0          8         454.0       220.0    4354   
           Usa     14.0          8         440.0       215.0    4312   
           Usa     14.0          8         455.0       225.0    4425   
           Usa     15.0          8         390.0       190.0    3850   
           Usa     15.0          8         383.0       170.0    3563   
           Usa     14.0          8         340.0       160.0    3609   
           Usa     15.0          8         400.0       150.0    3761   
           Usa     14.0          8         455.0       225.0    3086   
           Usa     22.0          6         198.0        95.0    2833   
           Usa     18.0          6         199.0        97.0    2774   
           Usa     21.0          6         200.0        85.0    2587   
           Usa     21.0          6         199.0        90.0    2648   
           Usa     10.0          8         360.0       215.0    4615   
           Usa     10.0          8         307.0       200.0    4376   
           Usa     11.0          8         318.0       210.0    4382   
           Usa      9.0          8         304.0       193.0    4732   
71         Europe  28.0          4         116.0        90.0    2123   
...                 ...        ...           ...         ...     ...   
82         Europe  44.0          4          97.0        52.0    2130   
           Japan   37.0          4          91.0        68.0    2025   
           Japan   31.0          4          91.0        68.0    1970   
           Japan   36.0          4         120.0        88.0    2160   
           Japan   36.0          4         107.0        75.0    2205   
           Japan   34.0          4         108.0        70.0    2245   
           Japan   38.0          4          91.0        67.0    1965   
           Japan   32.0          4          91.0        67.0    1965   
           Japan   38.0          4          91.0        67.0    1995   
           Japan   32.0          4         144.0        96.0    2665   
           Usa     28.0          4         112.0        88.0    2605   
           Usa     27.0          4         112.0        88.0    2640   
           Usa     34.0          4         112.0        88.0    2395   
           Usa     31.0          4         112.0        85.0    2575   
           Usa     29.0          4         135.0        84.0    2525   
           Usa     27.0          4         151.0        90.0    2735   
           Usa     24.0          4         140.0        92.0    2865   
           Usa     23.0          4         151.0         NaN    3035   
           Usa     38.0          4         105.0        63.0    2125   
           Usa     36.0          4          98.0        70.0    2125   
           Usa     25.0          6         181.0       110.0    2945   
           Usa     38.0          6         262.0        85.0    3015   
           Usa     26.0          4         156.0        92.0

The name of the second car is ... peugeot 504!

64. __Select__ all cars with the "model_year" __75__!

In [117]:
cars.loc[75]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
origin                                                                    
Europe  25.0          4          90.0        71.0    2223          16.5   
Europe  29.0          4          90.0        70.0    1937          14.0   
Europe  23.0          4         115.0        95.0    2694          15.0   
Europe  23.0          4         120.0        88.0    2957          17.0   
Europe  22.0          4         121.0        98.0    2945          14.5   
Europe  25.0          4         121.0       115.0    2671          13.5   
Japan   29.0          4          97.0        75.0    2171          16.0   
Japan   24.0          4         134.0        96.0    2702          13.5   
Japan   24.0          4         119.0        97.0    2545          17.0   
Japan   33.0          4          91.0        53.0    1795          17.5   
Usa     19.0          6         225.0        95.0    3264          16.0   
Usa     18.0          6         250.0       105.0    3459          16.0   
Usa     15.0          6         250.0        72.0    3432          21.0   
Usa     15.0          6         250.0        72.0    3158          19.5   
Usa     16.0          8         400.0       170.0    4668          11.5   
Usa     15.0          8         350.0       145.0    4440          14.0   
Usa     16.0          8         318.0       150.0    4498          14.5   
Usa     14.0          8         351.0       148.0    4657          13.5   
Usa     17.0          6         231.0       110.0    3907          21.0   
Usa     16.0          6         250.0       105.0    3897          18.5   
Usa     15.0          6         258.0       110.0    3730          19.0   
Usa     18.0          6         225.0        95.0    3785          19.0   
Usa     21.0          6         231.0       110.0    3039          15.0   
Usa     20.0          8         262.0       110.0    3221          13.5   
Usa     13.0          8         302.0       129.0    3169          12.0   
Usa     23.0          4         140.0        83.0    2639          17.0   
Usa     20.0          6         232.0       100.0    2914          16.0   
Usa     23.0          4         140.0        78.0    2592          18.5   
Usa     18.0          6         171.0        97.0    2984          14.5   
Usa     19.0          6         232.0        90.0    3211          17.0   

                             name  
origin                             
Europe                  vw dasher  
Europe                  vw rabbit  
Europe                 audi 100ls  
Europe                peugeot 504  
Europe                volvo 244dl  
Europe                  saab 99le  
Japan              toyota corolla  
Japan               toyota corona  
Japan                  datsun 710  
Japan            honda civic cvcc  
Usa       plymouth valiant custom  
Usa                chevrolet nova  
Usa               mercury monarch  
Usa                 ford maverick  
Usa              pontiac catalina  
Usa             chevrolet bel air  
Usa           plymouth grand fury  
Usa                      ford ltd  
Usa                 buick century  
Usa     chevrolet chevelle malibu  
Usa                   amc matador  
Usa                 plymouth fury  
Usa                 buick skyhawk  
Usa           chevrolet monza 2+2  
Usa               ford mustang ii  
Usa                    ford pinto  
Usa                   amc gremlin  
Usa                 pontiac astro  
Usa                    ford pinto  
Usa                     amc pacer

65. __Select__ all cars with the "model_year" __76__ from __japan__! __How many__ cars do we get?

In [125]:
cars.loc[(76,'Japan'),'name']

model_year  origin
76          Japan        honda civic
            Japan       datsun b-210
            Japan     toyota corolla
            Japan     toyota mark ii
Name: name, dtype: object

There are ... 4 cars!

66. __Select__ all cars with the "model_year" __77__ and __79__  from __japan__! And only the columns __"mpg"__ and __"name"__! What is the __mpg__ of __datsun 210__?

In [127]:
cars.loc[([77,79],['Japan','Usa']),['mpg','name']]

mpg                                  name
model_year origin                                            
77         Japan   31.5                     honda accord cvcc
           Japan   33.5                 datsun f-10 hatchback
           Japan   26.0               toyota corolla liftback
           Japan   30.0                             subaru dl
           Japan   22.0                            datsun 810
           Japan   21.5                            mazda rx-4
           Usa     30.0               buick opel isuzu deluxe
           Usa     25.5                     plymouth arrow gs
           Usa     17.5             chevrolet caprice classic
           Usa     17.0            oldsmobile cutlass supreme
           Usa     15.5                 dodge monaco brougham
           Usa     15.0               mercury cougar brougham
           Usa     17.5                    chevrolet concours
           Usa     20.5                         buick skylark
           Usa     19.0                plymouth volare custom
           Usa     18.5                          ford granada
           Usa     16.0                 pontiac grand prix lj
           Usa     15.5          chevrolet monte carlo landau
           Usa     15.5                      chrysler cordoba
           Usa     16.0                      ford thunderbird
           Usa     24.5                 pontiac sunbird coupe
           Usa     25.5                   ford mustang ii 2+2
           Usa     30.5                    chevrolet chevette
           Usa     33.5                        dodge colt m/m
79         Japan   34.1                      mazda glc deluxe
           Japan   31.8                            datsun 210
           Usa     21.5                     pontiac lemans v6
           Usa     19.8                      mercury zephyr 6
           Usa     22.3                       ford fairmont 4
           Usa     20.2                      amc concord dl 6
           Usa     20.6                         dodge aspen 6
           Usa     17.0             chevrolet caprice classic
           Usa     17.6                       ford ltd landau
           Usa     16.5                 mercury grand marquis
           Usa     18.2                       dodge st. regis
           Usa     16.9               buick estate wagon (sw)
           Usa     15.5              ford country squire (sw)
           Usa     19.2         chevrolet malibu classic (sw)
           Usa     18.5  chrysler lebaron town @ country (sw)
           Usa     35.7           dodge colt hatchback custom
           Usa     27.4                         amc spirit dl
           Usa     23.0                     cadillac eldorado
           Usa     23.9     oldsmobile cutlass salon brougham
           Usa     34.2                      plymouth horizon
           Usa     34.5                  plymouth horizon tc3
           Usa     28.4                 buick skylark limited
           Usa     28.8                    chevrolet citation
           Usa     26.8             oldsmobile omega brougham
           Usa     33.5                       pontiac phoenix

The mpg of datsun 210 is ... 31.8!

Multiindex slicing

In [86]:
cars.loc[(slice(74,75),slice('europe','usa')),['name','weight']]b

name  weight
model_year origin                                           
74         europe                           audi fox    2219
           europe                          vw dasher    1963
           europe                         opel manta    2300
           europe                           fiat 128    2108
           europe                        fiat 124 tc    2246
           europe                          fiat x1.9    2000
           japan                         datsun b210    1950
           japan                 toyota corolla 1200    1836
           japan                       toyota corona    1649
           japan                          datsun 710    2003
           japan                         honda civic    2489
           japan                              subaru    2391
           usa                       plymouth duster    3102
           usa                         ford maverick    2875
           usa                            amc hornet    2901
           usa                        chevrolet nova    3336
           usa                            ford pinto    2451
           usa                        chevrolet vega    2542
           usa     chevrolet chevelle malibu classic    3781
           usa                           amc matador    3632
           usa            plymouth satellite sebring    3613
           usa                      ford gran torino    4141
           usa              buick century luxus (sw)    4699
           usa             dodge coronet custom (sw)    4457
           usa                 ford gran torino (sw)    4638
           usa                      amc matador (sw)    4257
           usa                            dodge colt    2125
75         europe                          vw dasher    2223
           europe                          vw rabbit    1937
           europe                         audi 100ls    2694
           europe                        peugeot 504    2957
           europe                        volvo 244dl    2945
           europe                          saab 99le    2671
           japan                      toyota corolla    2171
           japan                       toyota corona    2702
           japan                          datsun 710    2545
           japan                    honda civic cvcc    1795
           usa               plymouth valiant custom    3264
           usa                        chevrolet nova    3459
           usa                       mercury monarch    3432
           usa                         ford maverick    3158
           usa                      pontiac catalina    4668
           usa                     chevrolet bel air    4440
           usa                   plymouth grand fury    4498
           usa                              ford ltd    4657
           usa                         buick century    3907
           usa             chevrolet chevelle malibu    3897
           usa                           amc matador    3730
           usa                         plymouth fury    3785
           usa                         buick skyhawk    3039
           usa                   chevrolet monza 2+2    3221
           usa                       ford mustang ii    3169
           usa                            ford pinto    2639
           usa                           amc gremlin    2914
           usa                         pontiac astro    2592
           usa                            ford pinto    2984
           usa                             amc pacer    3211

67. __Swap the levels__ of the MultiIndex and then __sort__ the new MultiIndex in __ascending__ order! __Reassign cars__! __Fill in the gaps__!

In [ ]:
cars = cars.---.---(ascending = True)

In [ ]:
# run the cell!
cars

68. __Reset__ the __index__ of __cars__ and create a RangeIndex! __Save__ the change!

In [89]:
cars.reset_index(inplace=True)

In [90]:
# run the cell!
cars

model_year  origin   mpg  cylinders  displacement  horsepower  weight  \
0            70  europe  26.0          4          97.0        46.0    1835   
1            70  europe  25.0          4         110.0        87.0    2672   
2            70  europe  24.0          4         107.0        90.0    2430   
3            70  europe  25.0          4         104.0        95.0    2375   
4            70  europe  26.0          4         121.0       113.0    2234   
5            70   japan  24.0          4         113.0        95.0    2372   
6            70   japan  27.0          4          97.0        88.0    2130   
7            70     usa  18.0          8         307.0       130.0    3504   
8            70     usa  15.0          8         350.0       165.0    3693   
9            70     usa  18.0          8         318.0       150.0    3436   
10           70     usa  16.0          8         304.0       150.0    3433   
11           70     usa  17.0          8         302.0       140.0    3449   
12           70     usa  15.0          8         429.0       198.0    4341   
13           70     usa  14.0          8         454.0       220.0    4354   
14           70     usa  14.0          8         440.0       215.0    4312   
15           70     usa  14.0          8         455.0       225.0    4425   
16           70     usa  15.0          8         390.0       190.0    3850   
17           70     usa  15.0          8         383.0       170.0    3563   
18           70     usa  14.0          8         340.0       160.0    3609   
19           70     usa  15.0          8         400.0       150.0    3761   
20           70     usa  14.0          8         455.0       225.0    3086   
21           70     usa  22.0          6         198.0        95.0    2833   
22           70     usa  18.0          6         199.0        97.0    2774   
23           70     usa  21.0          6         200.0        85.0    2587   
24           70     usa  21.0          6         199.0        90.0    2648   
25           70     usa  10.0          8         360.0       215.0    4615   
26           70     usa  10.0          8         307.0       200.0    4376   
27           70     usa  11.0          8         318.0       210.0    4382   
28           70     usa   9.0          8         304.0       193.0    4732   
29           71  europe  28.0          4         116.0        90.0    2123   
..          ...     ...   ...        ...           ...         ...     ...   
368          82  europe  44.0          4          97.0        52.0    2130   
369          82   japan  37.0          4          91.0        68.0    2025   
370          82   japan  31.0          4          91.0        68.0    1970   
371          82   japan  36.0          4         120.0        88.0    2160   
372          82   japan  36.0          4         107.0        75.0    2205   
373          82   japan  34.0          4         108.0        70.0    2245   
374          82   japan  38.0          4          91.0        67.0    1965   
375          82   japan  32.0          4          91.0        67.0    1965   
376          82   japan  38.0          4          91.0        67.0    1995   
377          82   japan  32.0          4         144.0        96.0    2665   
378          82     usa  28.0          4         112.0        88.0    2605   
379          82     usa  27.0          4         112.0        88.0    2640   
380          82     usa  34.0          4         112.0        88.0    2395   
381          82     usa  31.0          4         112.0        85.0    2575   
382          82     usa  29.0          4         135.0        84.0    2525   
383          82     usa  27.0          4         151.0        90.0    2735   
384          82     usa  24.0          4         140.0        92.0    2865   
385          82     usa  23.0          4         151.0         NaN    3035   
386          82     usa  38.0          4         105.0        63.0    2125   
387          82     usa  36

69. __Convert__ all elements in the __origin__ column to __titlecase__! __Overwrite__ the __origin__ column!

In [92]:
cars.origin = cars.origin.str.title()

In [93]:
# run the cell!
cars

model_year  origin   mpg  cylinders  displacement  horsepower  weight  \
0            70  Europe  26.0          4          97.0        46.0    1835   
1            70  Europe  25.0          4         110.0        87.0    2672   
2            70  Europe  24.0          4         107.0        90.0    2430   
3            70  Europe  25.0          4         104.0        95.0    2375   
4            70  Europe  26.0          4         121.0       113.0    2234   
5            70   Japan  24.0          4         113.0        95.0    2372   
6            70   Japan  27.0          4          97.0        88.0    2130   
7            70     Usa  18.0          8         307.0       130.0    3504   
8            70     Usa  15.0          8         350.0       165.0    3693   
9            70     Usa  18.0          8         318.0       150.0    3436   
10           70     Usa  16.0          8         304.0       150.0    3433   
11           70     Usa  17.0          8         302.0       140.0    3449   
12           70     Usa  15.0          8         429.0       198.0    4341   
13           70     Usa  14.0          8         454.0       220.0    4354   
14           70     Usa  14.0          8         440.0       215.0    4312   
15           70     Usa  14.0          8         455.0       225.0    4425   
16           70     Usa  15.0          8         390.0       190.0    3850   
17           70     Usa  15.0          8         383.0       170.0    3563   
18           70     Usa  14.0          8         340.0       160.0    3609   
19           70     Usa  15.0          8         400.0       150.0    3761   
20           70     Usa  14.0          8         455.0       225.0    3086   
21           70     Usa  22.0          6         198.0        95.0    2833   
22           70     Usa  18.0          6         199.0        97.0    2774   
23           70     Usa  21.0          6         200.0        85.0    2587   
24           70     Usa  21.0          6         199.0        90.0    2648   
25           70     Usa  10.0          8         360.0       215.0    4615   
26           70     Usa  10.0          8         307.0       200.0    4376   
27           70     Usa  11.0          8         318.0       210.0    4382   
28           70     Usa   9.0          8         304.0       193.0    4732   
29           71  Europe  28.0          4         116.0        90.0    2123   
..          ...     ...   ...        ...           ...         ...     ...   
368          82  Europe  44.0          4          97.0        52.0    2130   
369          82   Japan  37.0          4          91.0        68.0    2025   
370          82   Japan  31.0          4          91.0        68.0    1970   
371          82   Japan  36.0          4         120.0        88.0    2160   
372          82   Japan  36.0          4         107.0        75.0    2205   
373          82   Japan  34.0          4         108.0        70.0    2245   
374          82   Japan  38.0          4          91.0        67.0    1965   
375          82   Japan  32.0          4          91.0        67.0    1965   
376          82   Japan  38.0          4          91.0        67.0    1995   
377          82   Japan  32.0          4         144.0        96.0    2665   
378          82     Usa  28.0          4         112.0        88.0    2605   
379          82     Usa  27.0          4         112.0        88.0    2640   
380          82     Usa  34.0          4         112.0        88.0    2395   
381          82     Usa  31.0          4         112.0        85.0    2575   
382          82     Usa  29.0          4         135.0        84.0    2525   
383          82     Usa  27.0          4         151.0        90.0    2735   
384          82     Usa  24.0          4         140.0        92.0    2865   
385          82     Usa  23.0          4         151.0         NaN    3035   
386          82     Usa  38.0          4         105.0        63.0    2125   
387          82     Usa  36

Now, we want to __extract the manufacturer names__ from the column __"name"__. <br>
70. __Split__ the strings in the column __"name"__ on the very __first whitespace__ and __expand__ the splitted strings into __separate columns__! __Save__ new DataFrame in the variable __split__! __Fill in the gaps!__

In [96]:
split = cars.name.str.split(pat = ' ',n=1, expand = True)

__Inspect__!

In [97]:
# run the cell!
split

0                       1
0            vw       1131 deluxe sedan
1       peugeot                     504
2          audi                  100 ls
3          saab                     99e
4           bmw                    2002
5        toyota          corona mark ii
6        datsun                   pl510
7     chevrolet         chevelle malibu
8         buick             skylark 320
9      plymouth               satellite
10          amc               rebel sst
11         ford                  torino
12         ford             galaxie 500
13    chevrolet                  impala
14     plymouth                fury iii
15      pontiac                catalina
16          amc          ambassador dpl
17        dodge           challenger se
18     plymouth               'cuda 340
19    chevrolet             monte carlo
20        buick       estate wagon (sw)
21     plymouth                  duster
22          amc                  hornet
23         ford                maverick
24          amc                 gremlin
25         ford                    f250
26    chevrolet                     c20
27        dodge                    d200
28           ih                   1200d
29         opel                    1900
..          ...                     ...
368          vw                  pickup
369       mazda            glc custom l
370       mazda              glc custom
371      nissan               stanza xe
372       honda                  accord
373      toyota                 corolla
374       honda                   civic
375       honda            civic (auto)
376      datsun                  310 gx
377      toyota               celica gt
378   chevrolet                cavalier
379   chevrolet          cavalier wagon
380   chevrolet         cavalier 2-door
381     pontiac      j2000 se hatchback
382       dodge                aries se
383     pontiac                 phoenix
384        ford         fairmont futura
385         amc              concord dl
386    plymouth           horizon miser
387     mercury                  lynx l
388       buick         century limited
389  oldsmobile  cutlass ciera (diesel)
390    chrysler       lebaron medallion
391        ford               granada l
392       dodge             charger 2.2
393   chevrolet                  camaro
394        ford              mustang gl
395       dodge                 rampage
396        ford                  ranger
397   chevrolet                    s-10

[398 rows x 2 columns]

71. __Select__ the __first column__ and determine the __frequency of unique values__. __Fill in the gaps__! Which manufacturer is __most frequent__?

In [98]:
split[0].value_counts()

ford             51
chevrolet        47
plymouth         31
amc              28
dodge            28
toyota           26
datsun           23
vw               22
buick            17
pontiac          16
honda            13
mercury          12
mazda            12
oldsmobile       10
fiat              8
peugeot           8
audi              7
volvo             6
chrysler          6
renault           5
saab              4
opel              4
subaru            4
mercedes-benz     3
bmw               2
cadillac          2
triumph           1
ih                1
nissan            1
Name: 0, dtype: int64

There are 51 cars from the manufacturer ... Ford!

72. Let´s __doublecheck__ the result from above. Check __how many__ car names in the column "names" __contain "ford"__!

In [111]:
cars[cars.name.str.contains('ford')]['name']

11                  ford torino
12             ford galaxie 500
23                ford maverick
25                    ford f250
38                   ford pinto
42              ford torino 500
46             ford galaxie 500
49     ford country squire (sw)
54                 ford mustang
69          ford pinto runabout
73             ford galaxie 500
81        ford gran torino (sw)
83              ford pinto (sw)
99             ford gran torino
103                    ford ltd
111               ford maverick
114                ford country
119                  ford pinto
138               ford maverick
141                  ford pinto
146            ford gran torino
149       ford gran torino (sw)
165               ford maverick
169                    ford ltd
176             ford mustang ii
177                  ford pinto
180                  ford pinto
199            ford gran torino
202               ford maverick
207           ford granada ghia
210                  ford pinto
214     

It´s ... again 51! It seems that the very first word in the column "name" is always the manufacturer name.

# Well Done!

------------------------------

# Hints (Spoiler!)

61. apply() method, maximum - minimum

62. set_index() method, pass list with column labels

63. sort_index() method

64. pass outer index labels to loc operator

65. cars.loc[(outer index labels, inner index labels)]

66. cars.loc[(outer index labels, inner index labels), column labels]

67. methods swaplevel() and sort_index()

68. reset_index() method

69. string(str)-method title()

70. string(str)-method split(), n = 1, 

71. value_counts() method 

72. string(str)-method contains()